File: 2021/elixir/apps/day2/lib/part1.ex
defmodule Day2Part1 do
  @moduledoc """
  Day2 - Part1
  """

  defmodule SubmarineState do
    defstruct horizontal: 0, depth: 0
  end

  def solve(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.reduce(%SubmarineState{}, &dive/2)
    |> then(fn %SubmarineState{horizontal: h, depth: d} -> h * d end)
  end

  defp dive("forward " <> value, %SubmarineState{horizontal: h, depth: d}) do
    %SubmarineState{horizontal: h + String.to_integer(value), depth: d}
  end

  defp dive("up " <> value, %SubmarineState{horizontal: h, depth: d}) do
    %SubmarineState{horizontal: h, depth: d - String.to_integer(value)}
  end

  defp dive("down " <> value, %SubmarineState{horizontal: h, depth: d}) do
    %SubmarineState{horizontal: h, depth: d + String.to_integer(value)}
  end
end

File: 2021/go/day2/part1/part1.go
package part1

import (
	"bufio"
	"strconv"
	"strings"
)

type Command struct {
	Operation string
	Value     int
}

func Solve(input string) string {
	scanner := bufio.NewScanner(strings.NewReader(input))
	var commands []*Command
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " ")
		v, _ := strconv.Atoi(s[1])
		commands = append(commands, &Command{
			Operation: s[0],
			Value:     v,
		})
	}
	return strconv.Itoa(dive(commands))
}

func dive(commands []*Command) int {
	horizontal, depth := 0, 0
	for _, c := range commands {
		switch c.Operation {
		case "forward":
			horizontal += c.Value
		case "up":
			depth -= c.Value
		case "down":
			depth += c.Value
		}
	}
	return horizontal * depth
}
